### ISSUE TITLE
User can unpin a report and mark it as unread/read simultaneously.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#26917]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Right click on any report in LHS
2. Unpin/pin & Mark as read/unread quickly
## Expected Result: 
User should be able to select one at a time.
## Actual Result: 
User is able to select both at a time.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.65-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/b81892be-7e5b-491f-bf6b-69535b9df10a

https://github.com/Expensify/App/assets/43996225/11cf2559-d42c-46c6-8e50-0300f97641bb

**Expensify/Expensify Issue URL:**
**Issue reported by:** @krishna2323 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1692160684023549

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by akinwale** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User can unpin a report and mark it as unread/read simultaneously

### What is the root cause of that problem?

### What changes do you think we should make in order to solve the problem?


### What alternative solutions did you explore? (Optional)

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User can unpin a report and mark it as unread/read simultaneously.

### What is the root cause of that problem?
- Currently, when clicking on the "Mark as unread" (or "Mark as read") button in the context menu, it will delay for 800ms before hiding the context menu, like below: 
https://github.com/Expensify/App/blob/f68755cf202eb56b0bead752ddba4d037a57032e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L254
- The hideContextMenu with the first params called "shouldDelay":
https://github.com/Expensify/App/blob/f68755cf202eb56b0bead752ddba4d037a57032e/src/pages/home/report/ContextMenu/ReportActionContextMenu.js#L11-L32

### What changes do you think we should make in order to solve the problem?
- We clicking on "Mark as unread" button, we can call the hideContextMenu with shouldDelay param is false like:
```
   hideContextMenu(false, ReportActionComposeFocusManager.focus);
```

### What alternative solutions did you explore? (Optional)
- We can just disable the others button after clicking on "Mark as unread" button to make sure that others' logic is not broken by creating a state disableClickOnContextMenuItem in PopoverReportActionContextMenu that used to decide if we should disable context button item or not
- The hideContextMenu function will become: 
```
function hideContextMenu(shouldDelay, onHideCallback = () => {}) {
  ...
    contextMenuRef.current.disableClickOnContextMenuItem();
    setTimeout(() => {
        if (contextMenuRef.current.instanceID !== instanceID) {
            return;
        }
        contextMenuRef.current.hideContextMenu(onHideCallback);
    }, 800);
}
```
- In there, the disableClickOnContextMenuItem is:
```
  disableClickOnContextMenuItem() {
        this.setState({
            disableClickOnContextMenuItem: true,
        });
    }
```
### Result
- Solution 1: [Screencast from 24-08-2023 22:59:14.webm](https://github.com/Expensify/App/assets/141406735/d38d72c2-0517-4586-9e0e-de20042b0407)
- Solution 2: [Screencast from 25-08-2023 02:26:25.webm](https://github.com/Expensify/App/assets/141406735/01413fad-0ebc-4400-b163-183bb50828d7)

**Comment by DylanDylann** [comment]
@parasharrajat I think the expected is: User cannot click two items at once

**Comment by DylanDylann** [comment]
@parasharrajat Yeah, I mean that once the ContextMenu is open, user can click on  one item

**Comment by abekkala** [comment]
It was decided in the Slack discussion  that we should ensure the modal is closed when a context item is selected. Rather than staying open after the first selection

**Comment by abekkala** [comment]
@abdulrahuman5196 have you had a chance to review the proposals that have been provided?

**Comment by abekkala** [comment]
@abdulrahuman5196 have you reviewed the proposals for the fix?

**Comment by DylanDylann** [comment]
@abekkala  @abdulrahuman5196 Currently, we are using the animation when selecting Mark as unread or Pin so I think we should not close the modal immediately after selecting item

[Screencast from 15-09-2023 11:57:21.webm](https://github.com/Expensify/App/assets/141406735/7c35d980-2865-419c-bd1e-8cd52376d3b6)

**Comment by akinwale** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/26917#issuecomment-1709213430)

**Comment by DylanDylann** [comment]
@akinwale if you have a significant change to your proposal, please help add a new comment, not edit your original proposal. It will help the reviewer. Thanks!

**Comment by akinwale** [comment]
> @akinwale if you have a significant change to your proposal, please help add a new comment, not edit your original proposal. It will help the reviewer. Thanks!

https://github.com/Expensify/App/issues/26917#issuecomment-1722105043

**Comment by DylanDylann** [comment]
no. I think you need to write a new comment that related to this change. https://github.com/Expensify/App/issues/26431#issuecomment-1715348314

**Comment by akinwale** [comment]
> no. I think you need to write a new comment that related to this change. [#26431 (comment)](https://github.com/Expensify/App/issues/26431#issuecomment-1715348314)

Not necessarily. I'm following the contributing guide. Besides, the C+ has not reviewed the proposals yet.

**Comment by DylanDylann** [comment]
@abdulrahuman5196 The  2nd @akinwale `s proposal is similar to mine, which use a variable to keep track that whether the item is selected or not. Also, it is just edited after my proposal, and after my comment https://github.com/Expensify/App/issues/26917#issuecomment-1720665715 to indicate that we are using the animation after selecting the item

**Comment by abekkala** [comment]
@abdulrahuman5196 can you review the proposal and comments to get this one moving forward?

**Comment by abekkala** [comment]
@abdulrahuman5196 is the proposal above acceptable? or do you have comments?
If it's not a viable fix, please let me know ASAP and I can get someone else on this one to get fixed right away

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User can unpin a report and mark it as unread/read simultaneously.

### What is the root cause of that problem?
- Currently, when clicking on the "Mark as unread" (or "Mark as read") button in the context menu, it will delay for 800ms before hiding the context menu, like below: 
https://github.com/Expensify/App/blob/f68755cf202eb56b0bead752ddba4d037a57032e/src/pages/home/report/ContextMenu/ContextMenuActions.js#L254
- The hideContextMenu with the first params called "shouldDelay":
https://github.com/Expensify/App/blob/f68755cf202eb56b0bead752ddba4d037a57032e/src/pages/home/report/ContextMenu/ReportActionContextMenu.js#L11-L32

### What changes do you think we should make in order to solve the problem?
- We clicking on "Mark as unread" button, we can call the hideContextMenu with shouldDelay param is false like:
```
   hideContextMenu(false, ReportActionComposeFocusManager.focus);
```

### What alternative solutions did you explore? (Optional)
- We can just disable the others button after clicking on "Mark as unread" button to make sure that others' logic is not broken by creating a state disableClickOnContextMenuItem in PopoverReportActionContextMenu that used to decide if we should disable context button item or not
- The hideContextMenu function will become: 
```
function hideContextMenu(shouldDelay, onHideCallback = () => {}) {
  ...
    contextMenuRef.current.disableClickOnContextMenuItem();
    setTimeout(() => {
        if (contextMenuRef.current.instanceID !== instanceID) {
            return;
        }
        contextMenuRef.current.hideContextMenu(onHideCallback);
    }, 800);
}
```
- In there, the disableClickOnContextMenuItem is:
```
  disableClickOnContextMenuItem() {
        this.setState({
            disableClickOnContextMenuItem: true,
        });
    }
```
### Result
- Solution 1: [Screencast from 24-08-2023 22:59:14.webm](https://github.com/Expensify/App/assets/141406735/d38d72c2-0517-4586-9e0e-de20042b0407)
- Solution 2: [Screencast from 25-08-2023 02:26:25.webm](https://github.com/Expensify/App/assets/141406735/01413fad-0ebc-4400-b163-183bb50828d7)

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
@abekkala Sorry for the delay.

>Currently, we are using the animation when selecting Mark as unread or Pin so I think we should not close the modal immediately after selecting item

I agree on this. We can't close immediately.

What should be the expected behaviour here? Do we make the other options not clickable?

**Comment by DylanDylann** [comment]
@abdulrahuman5196 my proposal will make other options not clickable https://github.com/Expensify/App/issues/26917#issuecomment-1709379934

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 my proposal will make other options not clickable [#26917 (comment)](https://github.com/Expensify/App/issues/26917#issuecomment-1709379934)

But I am doubtful if we want it. Because if its not clickable how can the UI be? It would be better to confirm what exact behaviour we want? Like I have asked here https://github.com/Expensify/App/issues/26917#issuecomment-1739308258

cc: @abekkala

**Comment by DylanDylann** [comment]
@abdulrahuman5196 Yeah we need to confirm. In my opinion, my proposal will disable other options, and when hovering over these options, it will display as {cursor: not-allowed} and the color will change to any "none-actvie" color

**Comment by abekkala** [comment]
@Julesssss pulling you in here as you were on the Slack conversation. 

Should the modal close and not keep open to click back and forth both options?

**Comment by Julesssss** [comment]
> Currently, we are using the animation when selecting Mark as unread or Pin so I think we should not close the modal immediately after selecting item

I don't think the animation is necessary tbh. The user doesn't need to see the state change to understand that it has occurred, plus it is causing this bug and looks like UI lag. Having said that, if the animation is desired then let's just do nothing here. 

Let's ask @Expensify/design for their thoughts.

**Comment by shawnborton** [comment]
I'm actually curious why we think this is a bug? I mean if the menu is open, and the options are selectable, why can't the user select either one as quickly as they want? 

I mean obviously this is an edge case and we try to close the popover menu after one is selected. But I don't know that this is a "bug" personally.

**Comment by DylanDylann** [comment]
@shawnborton You can check this one https://github.com/Expensify/App/issues/24261. In that, we also prevent users from selecting two emojis at once

**Comment by shawnborton** [comment]
Sure, but this is different - these aren't the same action (adding an emoji) - they are two distinct actions.

**Comment by abekkala** [comment]
Ok, so it sounds like initially in the slack thread we decided this was something to "fix" and now after further reflection we decided this is not a bug. @Julesssss we good to just close this then?
